components/script/dom/notification.rs

```diff
@@ -0,0 +1,444 @@
+    fn new_inherited(
+        global: &GlobalScope,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: Option<ImmutableOrigin>,
+        base_url: Option<ServoUrl>,
+        fallback_timestamp: u64,
+    ) -> Fallible<Self> {
+        // If options["silent"] is true and options["vibrate"] exists, then throw a TypeError.
+        if options.silent.is_some() && options.vibrate.is_some() {
+            return Err(Error::Type(
+                "Can't specify vibration patterns when setting notification to silent.".to_string(),
+            ));
+        }
+        // If options["renotify"] is true and options["tag"] is the empty string, then throw a TypeError.
+        if options.renotify && options.tag.is_empty() {
+            return Err(Error::Type(
+                "tag must be set to renotify as an existing notification.".to_string(),
+            ));
+        }
+        // Set notification’s data to StructuredSerializeForStorage(options["data"]).
+        let data = Heap::boxed(options.data.get());
+        let title = title.clone();
+        let dir = options.dir;
+        let lang = options.lang.clone();
+        // FIXME: not sure which Origin should be used
+        let origin = origin.unwrap_or(ImmutableOrigin::new_opaque());
+        let body = options.body.clone();
+        let tag = options.tag.clone();
+        let image = options.image.as_ref().and_then(|image_url| {
+            ServoUrl::parse_with_base(base_url.as_ref(), image_url)
+                .ok()
+                .map(|url| USVString::from(url.to_string()))
+        });
+        let icon = options.icon.as_ref().and_then(|icon_url| {
+            ServoUrl::parse_with_base(base_url.as_ref(), icon_url)
+                .ok()
+                .map(|url| USVString::from(url.to_string()))
+        });
+        let badge = options.badge.as_ref().and_then(|badge_url| {
+            ServoUrl::parse_with_base(base_url.as_ref(), badge_url)
+                .ok()
+                .map(|url| USVString::from(url.to_string()))
+        });
+        let vibration_pattern = match &options.vibrate {
+            Some(pattern) => validate_and_normalize_vibration_pattern(pattern),
+            None => Vec::new(),
+        };
+        let timestamp = options.timestamp.unwrap_or(fallback_timestamp);
+        let renotify = options.renotify;
+        let silent = options.silent;
+        let require_interaction = options.requireInteraction;
+
+        let mut actions: Vec<NotificationAction> = Vec::new();
+        let max_actions = Notification::MaxActions(global);
+        for (idx, action) in options.actions.iter().enumerate() {
+            if idx >= max_actions as usize {
+                break;
+            }
+            actions.push(NotificationAction {
+                action: action.action.clone(), // Spec is using `name`, however in WebIDL it use `action`.
```

Comment:
Here you are confusing the public `NotificationAction` with the private https://notifications.spec.whatwg.org/#action